typing/jkind_types.mli

```diff
@@ -106,53 +106,61 @@ module Layout : sig
   end
 end
 
-module Baggage : sig
+module With_bounds : sig
+  module Type_info : sig
+    type +'type_expr t =
+      { type_expr : 'type_expr;
+        modality : Mode.Modality.Value.Const.t;
+        nullability : bool
+      }
+
+    val create :
+      type_expr:'type_expr ->
+      modality:Mode.Modality.Value.Const.t ->
+      deep_only:bool ->
+      'type_expr t
+  end
+
   type (+'type_expr, 'd) t =
-    | No_baggage : ('type_expr, 'l * 'r) t
+    | No_with_bounds : ('type_expr, 'l * 'r) t
     (* There must always be at least one type. *)
-    | Baggage :
-        'type_expr * 'type_expr list
+    | With_bounds :
+        'type_expr Type_info.t * 'type_expr Type_info.t list
         -> ('type_expr, 'l * Allowance.disallowed) t
 
-  val as_list : ('type_expr, 'l * 'r) t -> 'type_expr list
-
-  val has_baggage : ('type_expr, 'l * 'r) t -> bool
-end
-
-module Bound : sig
-  open Allowance
-
-  type (+'type_expr, 'd, 'a) t =
-    { modifier : 'a;
-      baggage : ('type_expr, 'd) Baggage.t
-    }
-    constraint 'd = 'l * 'r
+  include Allowance.Allow_disallow with type ('a, _, 'd) sided = ('a, 'd) t
 
-  val try_allow_l :
-    ('type_expr, 'l * 'r, 'a) t -> ('type_expr, allowed * 'r, 'a) t option
+  val as_list : ('type_expr, 'l * 'r) t -> 'type_expr Type_info.t list
 
-  val try_allow_r :
-    ('type_expr, 'l * 'r, 'a) t -> ('type_expr, 'l * allowed, 'a) t option
-end
+  val has_with_bounds : ('type_expr, 'l * 'r) t -> bool
 
-module Bounds : sig
-  include module type of Jkind_axis.Axis_collection (Bound)
+  val types_on_axis :
```

Comment:
Also this one. Does this ever need to be called on an r-kind?

Reply:
chased the implications of making this take an `(l * disallowed)` jkind and it ends up constraining round_up a bit. I think it's simpler to just make this return an empty list in that case.